File: bin/my_scripts/utils/format_split_lines.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Text.RegularExpressions;

class Program
{
    static void Main()
    {
        string path = "./"; // Current directory
        int totalChangeCount = 0;

        foreach (string filename in Directory.GetFiles(path, "*.c"))
        {
            Console.WriteLine($"Processing file: {filename}");
            int fileChangeCount = ProcessFile(filename);
            Console.WriteLine($"Done! Changes done: {fileChangeCount}");
            totalChangeCount += fileChangeCount;
        }

        Console.WriteLine($"Done processing files! Total changes done: {totalChangeCount}");
    }

    static bool CheckFirstLine(string line)
    {
        string tempLine = line.Trim();
        return (tempLine == "" || tempLine.StartsWith('/') || tempLine.StartsWith('*')) && !tempLine.StartsWith('#');
    }

    static bool CheckForComment(string line)
    {
        return line.Trim().StartsWith('/') || line.Trim().StartsWith("*");
    }

    static bool CheckForContent(string line)
    {
        return line.Trim() != "" && !CheckForComment(line) && !line.Trim().StartsWith('#');
    }

    static bool CheckForParen(string line)
    {
        return line.Trim().Contains('(');
    }

    static bool CheckForEndParen(string line)
    {
        return line.Trim().EndsWith(')');
    }

    static bool CheckForFuncStart(string line)
    {
        return line.Replace("\n", "").StartsWith("{");
    }

    static string GetFunctionName(string line)
    {
        Match match = Regex.Match(line, @"\b(\w+)\s*\(");
        if (match.Success)
        {
            return match.Groups[1].Value;
        }
        else
        {
            return null;
        }
    }

    static int ProcessFile(string filename)
    {
        int changeCounter = 0;
        List<string> lines = new List<string>(File.ReadAllLines(filename));
        List<string> newLines = new List<string>();

        int i = 0;
        while (i < lines.Count)
        {
            if (i < lines.Count - 2 && CheckFirstLine(lines[i]) && CheckForEndParen(lines[i + 1]) && CheckForFuncStart(lines[i + 2]))
            {
                newLines.Add(lines[i]);
                string funcName = GetFunctionName(lines[i + 1]);
                string returnType = lines[i + 1].Split(new string[] { funcName }, StringSplitOptions.None)[0];
                newLines.Add(returnType.Trim());
                newLines.Add(lines[i + 1].Substring(returnType.Length));
                newLines.Add(lines[i + 2]);
                i += 3;
                changeCounter++;
            }
            else if (i < lines.Count - 3 && CheckFirstLine(lines[i]) && CheckForParen(lines[i + 1]) && CheckForEndParen(lines[i + 2]) && CheckForFuncStart(lines[i + 3]))
            {
                newLines.Add(lines[i]);
                string funcName = GetFunctionName(lines[i + 1]);
                string returnType = lines[i + 1].Split(new string[] { funcName }, StringSplitOptions.None)[0];
                newLines.Add(returnType.Trim());
                newLines.Add(lines[i + 1].Substring(returnType.Length));
                newLines.Add(lines[i + 2]);
                newLines.Add(lines[i + 3]);
                i += 4;
                changeCounter++;
            }
            else if (i < lines.Count - 4 && CheckFirstLine(lines[i]) && !CheckForComment(lines[i + 1]) && CheckForParen(lines[i + 1]) && !CheckForComment(lines[i + 2]) && CheckForContent(lines[i + 2]) && CheckForEndParen(lines[i + 3]) && CheckForFuncStart(lines[i + 4]))
            {
                Console.WriteLine($"Found function with three lines of arguments, see: {lines[i + 1] + lines[i + 2] + lines[i + 3] + lines[i + 4]}");
                newLines.Add(lines[i]);
                i++;
            }
            else
            {
                newLines.Add(lines[i]);
                i++;
            }
        }

        File.WriteAllLines(filename, newLines);
        return changeCounter;
    }
}

File: .config/dwm_git/format_split_lines.py
import glob
import re

def get_function_name(line):
    # Regular expression to find the last word before parentheses
    match = re.search(r'\b(\w+)\s*\(', line)
    if match:
        return match.group(1)  # Returns the captured group, which is the function name
    else:
        return None  # No match found

def check_first_line(line):
    temp_line = line.strip()
    return (temp_line == '' or temp_line.startswith('/') or temp_line.startswith('*')) and not temp_line.startswith('#')

def check_for_comment(line):
    return line.strip().startswith('/') or line.strip().startswith("*")

def check_for_content(line):
    return line.strip() != '' and not check_for_comment(line) and not line.strip().startswith('#')

def check_for_paren(line):
    return '(' in line.strip()

def check_for_end_paren(line):
    return line.strip().endswith(')')

def check_for_func_start(line):
    return line.replace('\n', '').startswith('{')

def process_file(filename):
    change_counter = 0
    with open(filename, 'r') as file:
        lines = file.readlines()

    new_lines = []
    i = 0
    while i < len(lines):
        if i < len(lines) - 2 and check_first_line(lines[i]) and check_for_end_paren(lines[i+1]) and check_for_func_start(lines[i+2]):
            new_lines.append(lines[i])
            #paren_split = lines[i+1].split("(")
            #return_type = paren_split[0].split(" ")[len(paren_split[0].split(" "))-1]
            func_name = get_function_name(lines[i+1])
            return_type = lines[i+1].split(func_name)[0]
            new_lines.append(return_type.strip() + '\n')
            new_lines.append(lines[i+1].split(return_type, 1)[1])
            new_lines.append(lines[i+2])
            i += 3
            change_counter += 1
        elif i < len(lines) - 3 and check_first_line(lines[i]) and check_for_paren(lines[i+1]) and check_for_end_paren(lines[i+2]) and check_for_func_start(lines[i+3]):
            new_lines.append(lines[i])
            func_name = get_function_name(lines[i+1])
            return_type = lines[i+1].split(func_name)[0]
            new_lines.append(return_type.strip() + '\n')
            new_lines.append(lines[i+1].split(return_type, 1)[1])
            new_lines.append(lines[i+2])
            new_lines.append(lines[i+3])
            i += 4
            change_counter += 1
        elif i < len(lines) - 4 and check_first_line(lines[i]) and not check_for_comment(lines[i+1]) and check_for_paren(lines[i+1]) and not check_for_comment(lines[i+2]) and check_for_content(lines[i+2]) and check_for_end_paren(lines[i+3]) and check_for_func_start(lines[i+4]):
            print(f"Found function with three lines of arguments, see: {lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4]}")
            new_lines.append(lines[i])
            i += 1
        else:
            new_lines.append(lines[i])
            i += 1

    with open(filename, 'w') as file:
        file.writelines(new_lines)
    return change_counter

# Replace the path with the directory where your .c files are located
path = './' # Current directory
total_change_count = 0
for filename in glob.glob(path + '*.c'):
    print(f'Processing file: {filename}')
    file_change_count = process_file(filename)
    print(f"Done! Changes done: {file_change_count}")
    total_change_count += file_change_count

print(f"Done processing files! Total changes done: {total_change_count}")
